test/unittest/tests/block/test_block_api_misc.cpp

```diff
@@ -0,0 +1,242 @@
+    memcpy(WT_BLOCK_HEADER_BYTE(buf->mem), contents.c_str(), contents.length());
+}
+
+static void
+initialize_bm(std::shared_ptr<mock_session> &session, WT_BM *bm)
+{
+    // Build a mock session, this will automatically create a mock connection.
+    // Additionally, initialize a session implementation variable for easy access.
+    session = mock_session::build_test_mock_session();
+    WT_SESSION_IMPL *s = session->get_wt_session_impl();
+
+    REQUIRE(session->get_mock_connection()->setup_block_manager(s) == 0);
+    session->setup_block_manager_file_operations();
+
+    // Set up the block manager so that we can use its legal API methods.
+    __ut_bm_method_set(bm);
+
+    // Initialization steps for the block manager. We shouldn't need to touch the block checkpoint
+    // logic and this shows that WiredTiger has poor module separation.
+    auto path = std::filesystem::current_path();
+    std::string file_path(path.string() + "/test.wt");
+    REQUIRE(__wt_block_manager_create(s, file_path.c_str(), DEFAULT_BLOCK_SIZE) == 0);
+
+    config_parser cp({{"allocation_size", ALLOCATION_SIZE}, {"block_allocation", BLOCK_ALLOCATION},
+      {"os_cache_max", OS_CACHE_MAX}, {"os_cache_dirty_max", OS_CACHE_DIRTY_MAX},
+      {"access_pattern_hint", ACCESS_PATTERN}});
+    REQUIRE(__wt_block_open(s, file_path.c_str(), WT_TIERED_OBJECTID_NONE, cp.get_config_array(),
+              false, false, false, DEFAULT_BLOCK_SIZE, &bm->block) == 0);
+    REQUIRE(__wti_block_ckpt_init(s, &bm->block->live, nullptr) == 0);
+}
+
+static void
+check_bm_stats(WT_SESSION_IMPL *session, WT_BM *bm)
+{
+    WT_DSRC_STATS stats;
+
+    // Enable statistics on the connection to allow the statistic macros to update correctly.
+    S2C(session)->stat_flags = 1;
+    REQUIRE(bm->stat(bm, session, &stats) == 0);
+    CHECK(stats.allocation_size == bm->block->allocsize);
+    CHECK(stats.block_checkpoint_size == (int64_t)bm->block->live.ckpt_size);
+    CHECK(stats.block_magic == WT_BLOCK_MAGIC);
+    CHECK(stats.block_major == WT_BLOCK_MAJOR_VERSION);
+    CHECK(stats.block_minor == WT_BLOCK_MINOR_VERSION);
+    CHECK(stats.block_reuse_bytes == (int64_t)bm->block->live.avail.bytes);
+    CHECK(stats.block_size == bm->block->size);
+
+    // Disable statistics on the connection when finished so that the mock session destructor
+    // doesn't try to dereference invalid memory.
+    S2C(session)->stat_flags = 0;
+}
+
+// Test that the block manager's addr_string method produces the expected string representation.
+static void
+test_addr_string(WT_SESSION_IMPL *session, WT_BM *bm, wt_off_t pack_offset, uint32_t pack_size,
+  uint32_t pack_checksum, std::string expected_str)
+{
+    // Initialize a buffer.
+    WT_ITEM buf;
+    WT_CLEAR(buf);
+
+    // Generate an address cookie - technically, we shouldn't know about internal details of the
+    // address cookie, but this allows for more rigorous testing with different inputs.
+    uint8_t p[WT_BTREE_MAX_ADDR_COOKIE], *pp;
```

Comment:
I don't think we need pp here, and just use &p instead here

Reply:
&p is wrong.  Just use p. p is an array. p gives you a pointer to the [0] element of the array. That is why the assignment pp = p; works.

There are two patterns here.
* For an array: type data[size]; type *datap = data;
* For any other type: type data, *datap; datap = &data;

The array pattern doesn't make much sense since an array itself is treated as a pointer to the type. (I actually don't like the other one either.) The only reason to use the array pattern is if you also use the other pattern it hides whether data is an array and allows you to use datap whenever a pointer is needed without considering whether data is an array or not.

This pattern of type data, *datap; comes from struct types. For those you do need the & and the pattern works. For arrays the pattern does not work.

Reply:
thanks! that makes sense to me.